src/pynxtools/dataconverter/readers/multi/reader.py

```diff
@@ -0,0 +1,446 @@
+
+    Also takes care of the "!" notation, i.e., removes optional groups
+    if a required sub-element cannot be filled.
+    """
+
+    def try_convert(value: str) -> Union[str, float, int, bool]:
+        """
+        Try to convert the value to float, int or bool.
+        If not convertible returns the str itself.
+        """
+
+        if is_integer(value):
+            return int(value)
+        if is_number(value):
+            return float(value)
+        if is_boolean(value):
+            return to_bool(value)
+
+        return value
+
+    def parse_config_value(value: str) -> Tuple[str, Any]:
+        """
+        Separates the prefixes (denoted by "@") from the rest
+        of the value.
+
+        Parameters
+        ----------
+        value : str
+            Config dict value.
+
+        Returns
+        -------
+        Tuple[str, Any]
+            Tuple like (prefix, path).
+
+        """
+        # Regex pattern to match @prefix:some_string
+        pattern = r"!?(@\w+)(?::(.*))?"
+        prefixes = re.findall(pattern, value)
+        if not prefixes:
+            return ("", value)
+        return prefixes[0]
+
+    # Handle non-keyword values
+    if not isinstance(value, str) or "@" not in str(value):
+        if isinstance(value, str):
+            # Handle "!" notation for required fields in optional groups
+            value = value.lstrip("!")
+        new_entry_dict[key] = value
+        return
+
+    prefixes: List[Tuple[str, str]] = []
+
+    try:
+        # Safely evaluate the string to a list
+        list_value = ast.literal_eval(value)
+        prefixes = [parse_config_value(v) for v in list_value]
+    except (SyntaxError, ValueError):
+        prefixes = [parse_config_value(value)]
+
+    for prefix, path in prefixes:
+        if not prefix.startswith("@"):
+            new_entry_dict[key] = try_convert(path)
+            break
+        new_entry_dict[key] = callbacks.apply_special_key(prefix, key, path)
+
+        # We found a match. Stop resolving other prefixes.
+        if new_entry_dict[key] is not None:
+            break
+
+    if value.startswith("!") and new_entry_dict[key] is None:
+        group_to_delete = key.rsplit("/", 1)[0]
+        logger.info(
+            f"Main element {key} not provided. "
+            f"Removing the parent group {group_to_delete}."
+        )
+        optional_groups_to_remove.append(group_to_delete)
+        return
+
+    if prefixes and new_entry_dict[key] is None:
+        del new_entry_dict[key]
+        if not suppress_warning:
+            logger.warning(
+                f"Could not find value for key {key} with value {value}.\n"
+                f"Tried prefixes: {prefixes}."
+            )
+
+    # after filling, resolve links again:
+    if isinstance(new_entry_dict.get(key), str) and new_entry_dict[key].startswith(
+        "@link:"
+    ):
+        new_entry_dict[key] = {"link": new_entry_dict[key][6:]}
```

Comment:
There is this create_link_dict option in the json parser now. I don't understand how it works and what it's good for: Does it have to be considered here as well?

Reply:
No, it's just backword compatibility (see the other comment on the issue)